include/dr/shp/copy.hpp

```diff
@@ -91,29 +74,67 @@ OutputIt copy(InputIt first, InputIt last, OutputIt d_first) {
     std::advance(local_last, n_to_copy);
 
     auto remote_iter = segment.begin();
-    std::advance(remote_iter, local_id);
 
     auto event = shp::copy_async(first, local_last, remote_iter);
 
     events.push_back(event);
 
-    local_id = 0;
     segment_id++;
     std::advance(first, n_to_copy);
     total_copied += n_to_copy;
   }
 
-  for (auto &&event : events) {
-    event.wait();
-  }
+  sycl::queue q;
+  auto root_event = q.submit([=](auto &&h) { h.depends_on(events); });
+  root_event.wait();
 
   auto rv = d_first;
   std::advance(rv, total_copied);
   return rv;
 }
 
+template <std::forward_iterator InputIt, lib::distributed_iterator OutputIt>
+  requires is_syclmemcopyable<std::iter_value_t<InputIt>,
+                              std::iter_value_t<OutputIt>>
+cl::sycl::event copy_async(InputIt first, InputIt last, OutputIt d_first) {
+  auto segments = lib::ranges::segments(d_first);
+
+  std::size_t segment_id = 0;
+
+  std::vector<cl::sycl::event> events;
+
+  std::size_t total_copied = 0;
+
+  while (first != last) {
+    auto &&segment = segments[segment_id];
+    std::size_t n_in_segment = segment.size();
+
+    std::size_t n_to_copy =
+        std::min<size_t>(n_in_segment, std::distance(first, last));
+
+    auto local_last = first;
+    std::advance(local_last, n_to_copy);
+
+    auto remote_iter = segment.begin();
```

Comment:
`remote_iter` variable is used in one place only. There is no point in making this variable. Just pass `segment.begin()` into `shp::copy_async` call (here and in the function above)